forge-game/src/main/java/forge/game/player/Player.java

```diff
@@ -2766,12 +2767,34 @@ public void onMulliganned() {
     public List<Card> getCommanders() {
         return commanders;
     }
-    public void setCommanders(List<Card> commanders0) {
-        if (commanders0 == commanders) { return; }
-        commanders = commanders0;
+    public void setCommanders(List<Card> commanders) {
+        if (new HashSet<>(this.commanders).containsAll(commanders)) { return; }
```

Comment:
imo should that be more optimized?

Should this.commanders be a CardCollection to stop duplicates?

Reply:
That would make sense, yes. And I also realized it should probably be a test if the sets match exactly; if you have commanders A and B, setCommanders([A]) should remove B. I'll have another shot at this bit.

Reply:
Alright, new version should be safe from duplicates, and should skip over any commanders on the input list that the player already has as their commander. It also manages the commander effect internally rather than leaving that to the call site. I also added a `copyCommandersToSnapshot` method so the map from commanders to commander effects could be preserved when copying the game state, and moved the logic for the other commander maps in there too.